File: Python/Problems/ProblemSets/Problem1-2.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class LinkedList:
    def __init__(self):
        self.head = None

    def push(self, val):
        new_node = Node(val)
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while last.next is not None:
            last = last.next
        last.next = new_node
        new_node.prev = last

    def __str__(self):
        ret_str = "["
        temp = self.head
        while temp is not None:
            ret_str += str(temp.val) + ", "
            temp = temp.next
        ret_str = ret_str.rstrip(", ")
        ret_str += "]"
        return ret_str
    

    
    def count(self):
        counter = 0
        temp = self.head 
        while temp is not None:
            counter += 1
            temp = temp.next
        return counter
    
#[1,2,2,3,2,1]  =>2 
    
    def same_elements(self, val ):
        counter = 0
        temp = self.head
        while temp is not None:
            if temp.val == val:
                counter += 1  # 1, 2, 3
            temp = temp.next
        return counter
    




    
l = LinkedList()

for i in [1, 2, 2, 2, 3, 4]:
    l.push(i)

print(l)
print(l.count())

print(l)
print(l.same_elements(2))

File: Cpp/Problems/ProblemSets/Problem1-2.cpp

#include <iostream>
using namespace std;
// int main()

class Node
{
public:
    int val;
    Node *next;

    Node(int v)
    {
        val = v;
        next = nullptr;
    }
};

class LinkedList
{
public:
    Node *head;
    LinkedList()
    {
        head = nullptr;
    }

    void Push(int val)
    {
        Node *new_node = new Node(val);
        if (head == nullptr)
        {
            head = new_node;
            return;
        }

        Node *last = head;
        while (last->next != nullptr)
        {
            last = last->next;
        }

        last->next = new_node;
    }

    string toString()
    {
        string ret_str = "[";
        Node *temp = head;
        while (temp != nullptr)
        {
            ret_str += to_string(temp->val) + ", ";
            temp = temp->next;
        }

        if (ret_str.size() > 1)
        {
            ret_str.pop_back();
            ret_str.pop_back();
        }

        ret_str += "]";
        return ret_str;
    };

    int count_nodes()
    {
        int count = 0;
        Node *temp = head;
        while (temp != nullptr)
        {
            count++;
            temp = temp->next;
        }
        return count;
    }

    int same_elements(int val)
    {
        int count = 0;
        Node *temp = head;
        while (temp != nullptr)
        {
            if (temp->val == val)
            {
                count++;
            }
            temp = temp->next;
        }
        return count;
    }
};

int main()
{
    LinkedList l;
    int arr[] = {2, 3, 4, 5, 2, 8, 7, 7, 5};
    int size = sizeof(arr) / sizeof(arr[0]);

    for (int i = 0; i < size; i++)
    {
        l.Push(arr[i]);
    }
    cout << l.toString() << endl;
    cout << "Occurrences of 5: " << l.same_elements(5) << endl;

    return 0;
}